_lzbench/lzbench.h

```diff
@@ -212,15 +209,16 @@ static const compressor_desc_t comp_desc[LZBENCH_COMPRESSOR_COUNT] =
     { "zstd22LDM",  "1.4.9",       1,  22,   22,       0, lzbench_zstd_LDM_compress,   lzbench_zstd_decompress,       lzbench_zstd_LDM_init,   lzbench_zstd_deinit },
     { "zstd24LDM",  "1.4.9",       1,  22,   24,       0, lzbench_zstd_LDM_compress,   lzbench_zstd_decompress,       lzbench_zstd_LDM_init,   lzbench_zstd_deinit },
     { "nakamichi",  "okamigan",    0,   0,    0,       0, lzbench_nakamichi_compress,  lzbench_nakamichi_decompress,  NULL,                    NULL },
-    { "cudaMemcpy", "",            0,   0,    0,       0, lzbench_return_0,            lzbench_cuda_memcpy,           lzbench_cuda_init,       lzbench_cuda_deinit },
+#ifdef BENCH_HAS_CUDA
```

Comment:
I think this is unnecessary. because we already have：
#ifdef BENCH_HAS_CUDA
        char* lzbench_cuda_init(size_t insize, size_t, size_t);
        void lzbench_cuda_deinit(char* workmem);
        int64_t lzbench_cuda_memcpy(char *inbuf, size_t insize, char *outbuf, size_t outsize, size_t, size_t, char* workmem);
        int64_t lzbench_cuda_return_0(char *inbuf, size_t insize, char *outbuf, size_t outsize, size_t , size_t, char* );
#else
        #define lzbench_cuda_init NULL
        #define lzbench_cuda_deinit NULL
        #define lzbench_cuda_memcpy NULL
        #define lzbench_cuda_return_0 NULL
#endif
We should keep the original style, like BENCH_REMOVE_LZSSE.

Reply:
Personally, I find it more efficient, robust, *etc.* to not have the invalid entries in the first place, rather than including them and checking at runtime.

If you would prefer consistency, I can make the same change for all the others, too.

Reply:
> Personally, I find it more efficient, robust, _etc._ to not have the invalid entries in the first place, rather than including them and checking at runtime.
> 
> If you would prefer consistency, I can make the same change for all the others, too.

which part of efficiency have been improved?

Reply:
With the current approach, when a tool is disabled:
  - the `compressor_desc_t` array still has an entry for it
  - the tool is still accepted as a valid argument to `lzbench -e`
  - the loop at https://github.com/inikep/lzbench/blob/2b5fda842a0c935ebb20f14ed55aa5f417bba276/_lzbench/lzbench.cpp#L873-L882
    needs to check each entry to see if its `compress` is valid or `NULL`
  - similarly, `lzbench_test` needs to check each tool for validity, and otherwise skip it

For example, if lzsse is disabled (*e.g.*  by setting `DONT_BUILD_LZSSE=1`), it is still accepted as a valid tool, and then it is skipped at runtime:
```bash
$ ./lzbench -elzsse2 data
lzbench 1.8 (64-bit Linux)   Assembled by P.Skibinski
Compressor name         Compress. Decompress. Compr. size  Ratio Filename
memcpy                  28864 MB/s 28917 MB/s     1048576 100.00 data
done... (cIters=1 dIters=1 cTime=1.0 dTime=2.0 chunkSize=1706MB cSpeed=0MB)
```

If one changes the definition of `compressor_desc_t` to exclude the tools that are not compiled in, and adjusts its size automatically, then it's possible to remove the various checks from `lzbench.cpp`.
IMHO that would make the code marginally cleaner, and the runtime behaviour more consistent.

Reply:
Of course, that **is** a change in behaviour, so it all depends if the current behaviour is intended, whether `-etool` should silently ignore or fail, for a `tool` that is generally supported but conditionally disabled at compile time.